### ISSUE TITLE
Skipping Reason

### ISSUE THREAD

**Issue by jerop** [tektoncd/pipeline#4738]
### Feature Request

Surface the reason why a specific `Task` in a `Pipeline` was skipped.

Today, users only know that a `PipelineTask` was skipped, but they don't know for which exact reason. When users see a `PipelineTask` has been skipped without knowing the reason, they may get confused about the behavior of the `PipelineRun` e.g. https://github.com/tektoncd/pipeline/issues/4571 and [slack thread][slack].

There are many [reasons][reasons] why a `PipelineTask` could be skipped, including:
* at least one of its `when` expressions evaluated to false 
* at least one of its `Conditions` failed
* at least one of the `Results` it was consuming was missing
* at least one its parent `PipelineTasks` was skipped
* the `PipelineRun` was in stopping state
* the `PipelineRun` was gracefully cancelled
* the `PipelineRun` was gracefully stopped

### Use case

As a `Pipeline` user, I need to know why certain `Tasks` in my `Pipeline` have been skipped.

[skipped-tasks]: https://github.com/tektoncd/pipeline/blob/053833cb10f3829d5a366daa1f431b293dcf3285/pkg/apis/pipeline/v1beta1/pipelinerun_types.go#L466-L476
[issue]: https://github.com/tektoncd/pipeline/issues/4571
[slack]: https://tektoncd.slack.com/archives/CK3HBG7CM/p1642349040014100
[reasons]: https://github.com/tektoncd/pipeline/blob/053833cb10f3829d5a366daa1f431b293dcf3285/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L42-L62

**Comment by chitrangpatel** [comment]
Hi, I'm working on this issue in this [PR](https://github.com/tektoncd/pipeline/pull/4800). At this point, all the unit tests pass and after adding the `Reason` field where necessary. 

 This feature definitely tracks the Skipping Reason. My question is, how is it accessed by the users? Using the `tkn pipelinerun describe` cli, I can see the skipped step as shown below. Do I need to update the CLI to also display the `skipping reason` or was the intention to make it visible in some other way?
 ```
$ tkn pipelinerun describe pipelinerun-skip-task-run-task
Name:              pipelinerun-skip-task-run-task
Namespace:         default
Service Account:   default
Timeout:           1h0m0s
Labels:
 tekton.dev/pipeline=pipelinerun-skip-task-run-task

🌡️  Status

STARTED          DURATION   STATUS
27 minutes ago   5s         Succeeded(Completed)

🗂  Taskruns

 NAME                                             TASK NAME       STARTED          DURATION   STATUS
 ∙ pipelinerun-skip-task-run-task-run-this-task   run-this-task   27 minutes ago   5s         Succeeded

⏭️  Skipped Tasks

 NAME               REASON
 ∙ skip-this-task
```

**Comment by chitrangpatel** [comment]
Checking the status using the following command:
`kubectl get pipelinerun pipelinerun-skip-task-run-task -o yaml` results in: 
```
status:
  completionTime: "2022-04-27T14:31:37Z"
  conditions:
  - lastTransitionTime: "2022-04-27T14:31:37Z"
    message: 'Tasks Completed: 1 (Failed: 0, Cancelled 0), Skipped: 1'
    reason: Completed
    status: "True"
    type: Succeeded
  ...
  skippedTasks:
  - name: skip-this-task
    reason: WhenExpressionsSkip
    whenExpressions:
    - input: foo
      operator: in
      values:
      - foo1
  ...
```
Under `skippedTasks`, there is now a `reason: WhenExpressionsSkip` showing the reason why the test was skipped.

**Comment by chitrangpatel** [comment]
Currently, when we use the `tkn` CLI as shown below, `Skipped Tasks` only displays the `NAME`. Should it also show the `REASON` for skipping the task?

```
chitrang-macbookpro:cli chitrang$ ./tkn pipelinerun describe pipelinerun-skip-task-run-task
Name:              pipelinerun-skip-task-run-task
Namespace:         default
Service Account:   default
Timeout:           1h0m0s
Labels:
 tekton.dev/pipeline=pipelinerun-skip-task-run-task

🌡️  Status

STARTED          DURATION   STATUS
52 minutes ago   7s         Succeeded(Completed)

🗂  Taskruns

 NAME                                             TASK NAME       STARTED          DURATION   STATUS
 ∙ pipelinerun-skip-task-run-task-run-this-task   run-this-task   52 minutes ago   7s         Succeeded

⏭️  Skipped Tasks

 NAME              
 ∙ skip-this-task  
```

**Comment by jerop** [comment]
@chitrangpatel - yes, showing `Reason` in the CLI will be useful - we can ask the CLI team to take a look after the feature is released (cc @piyush-garg @vinamra28 @vdemeester)

**Comment by lbernick** [comment]
/close

this feature has been implemented in #4829. If we want to add this functionality to the CLI let's create an issue in that repo instead of tracking here.
